### PR TITLE
Adding Testing Suite and Tests

### PR DESCRIPTION
**Pull request by AbsentSemicolon**
Using Jasmine-Node for testing suite and istanbul to provide coverage which is fed into Codecov to show in readme. Also including dependencies banners and integration into Travis-CI

### CODE REVIEW

File: spec/logger.spec.js [hunk 1]
```diff
@@ -0,0 +1,57 @@
+describe("logger", function () {
+    "use strict";
+
+    var log, logger;
+
+    logger = require("../lib/logger");
```
Comment (fidian) [inline spec/logger.spec.js]: Should be in the beforeEach for it to be reliably reset.

File: spec/logger.spec.js [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+        });
+
+        it("prints to log from info", function () {
+            log.info("something");
+            expect(console.log).toHaveBeenCalledWith("something");
+        });
+
+        it("prints to error from error", function () {
+            log.error("something");
+            expect(console.error).toHaveBeenCalledWith("ERROR: something");
```
Comment (fidian) [inline spec/logger.spec.js]: Should expect that the logged string has "ERROR:" in front.

File: spec/logger.spec.js [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+        });
+
+        it("prints to error from warn", function () {
+            log.warn("something");
+            expect(console.error).toHaveBeenCalledWith("WARN: something");
+        });
+
+        it("prints to log from console", function () {
+            log.console("something");
+            expect(console.log.apply.mostRecentCall.args[1]).toEqual(["something"]);
```
Comment (fidian) [inline spec/logger.spec.js]: Doesn't it work if you `expect(console.log).toHaveBeenCalledWith("something")`?

File: spec/web-server.spec.js [hunk 4]
```diff
@@ -0,0 +1,259 @@
+describe("web-server", function () {
+    "use strict";
+
+    var fs, logger, restify, restMiddleware, webServer, WebServer;
+
+    WebServer = require("../lib/web-server");
```
Comment (fidian) [inline spec/web-server.spec.js]: Should move to the beforeEach.

File: spec/web-server.spec.js [hunk 5]
```diff
@@ -0,0 +1,259 @@
...
+        spyOn(fs, "readFileSync");
+        spyOn(logger, "console");
+        spyOn(logger, "debug");
+        spyOn(logger, "error");
+        spyOn(logger, "info");
+        spyOn(logger, "warn");
+        spyOn(restify, "createServer").andReturn({});
+    });
+
+    beforeEach(function () {
```
Comment (fidian) [inline spec/web-server.spec.js]: Can this get merged with the above?  If so, can you drop some of those extra variables so they are scoped to only the callback inside the beforeEach?

File: spec/web-server.spec.js [hunk 6]
```diff
@@ -0,0 +1,259 @@
...
+                return {};
+            }
+        };
+
+        restMiddleware = function () {
+
+        };
+
+        spyOn(fs, "readFileSync");
+        spyOn(logger, "console");
```
Comment (fidian) [inline spec/web-server.spec.js]: We should have a fake logger module that does the spying automatically for you.  `require("./mocks/logger-mock")` sort of thing.

File: spec/web-server.spec.js [hunk 7]
```diff
@@ -0,0 +1,259 @@
...
+        spyOn(restify, "createServer").andReturn({});
+    });
+
+    beforeEach(function () {
+        webServer = new WebServer(fs, logger, restify, restMiddleware);
+    });
+
+    describe("constructor and configure", function () {
+        it("should set up properties without config options then calls configure", function () {
+            expect(webServer.config).toEqual({
```
Comment (fidian) [inline spec/web-server.spec.js]: We should be testing the behavior, not the internal properties.  Mostly things like "did the instance get created" and "Is it set up for HTTPS automatically" or stuff.  We should not peek inside the webServer.config (as we discussed earlier today).

File: spec/web-server.spec.js [hunk 8]
```diff
@@ -0,0 +1,259 @@
...
+                handleUncaughtExceptions: true,
+                handleUpgrades: false,
+                httpsServerOptions: null,
+                name: "OpenToken API Test",
+                proxyProtocol: true,
+                spdy: true,
+                version: "2"
+            });
+
+            expect(fs.readFileSync).toHaveBeenCalledWith("./cert.crt");
```
Comment (fidian) [inline spec/web-server.spec.js]: Very fragile.  Ideally we would not need to mock out fs.

File: spec/logger.spec.js (hunk 3 above)
Comment (AbsentSemicolon) [inline spec/logger.spec.js]: If I don't spy on console.log.apply I can do expect(console.log).toHaveBeenCalledWith("something"). I'll go with this route.

File: spec/logger.spec.js (hunk 1 above)
Comment (AbsentSemicolon) [inline spec/logger.spec.js]: Have to leave variables out of before each so I can use them in tests, but the require I moved into beforeEach.

**Comment by fidian** [comment]
I have reviewed the commits and the end result looks just fine to me.  A great start for getting testing going.  Need to get confirmation that another person approves of my changes.

**Comment by AbsentSemicolon** [comment]
The changes not done by me look good and give a good basis for tests.

**Comment by fidian** [comment]
Squashing!

**Pull request merged**